File: server/system.js
"use strict"

const os = require("os")
const { getDockerInstance, isMock } = require("./docker")

const REPO = "SakithaSamarathunga33/vps"

// ── Version ───────────────────────────────────────────────────────────────────

let _versionCache = null
let _versionCacheAt = 0

function currentVersion() {
  // Injected at Docker build time; falls back to package.json
  return process.env.PULSENODE_VERSION || require("../package.json").version || "dev"
}

async function getVersionInfo() {
  const now = Date.now()
  if (_versionCache && now - _versionCacheAt < 3_600_000) return _versionCache

  const current = currentVersion()
  let latest = null, releaseUrl = null, changelog = null

  try {
    const res = await fetch(`https://api.github.com/repos/${REPO}/releases/latest`, {
      headers: { Accept: "application/vnd.github.v3+json", "User-Agent": "PulseNode" },
      signal: AbortSignal.timeout(6000),
    })
    if (res.ok) {
      const d = await res.json()
      latest     = d.tag_name?.replace(/^v/, "") || null
      releaseUrl = d.html_url || null
      changelog  = d.body     || null
    }
  } catch { /* network unavailable — return current only */ }

  const hasUpdate = latest ? _isNewer(latest, current) : false
  _versionCache   = { current, latest, hasUpdate, releaseUrl, changelog }
  _versionCacheAt = now
  return _versionCache
}

function _isNewer(a, b) {
  // returns true if a > b (a is newer)
  const pa = a.split(".").map(Number)
  const pb = b.split(".").map(Number)
  for (let i = 0; i < Math.max(pa.length, pb.length); i++) {
    const va = pa[i] ?? 0, vb = pb[i] ?? 0
    if (va > vb) return true
    if (va < vb) return false
  }
  return false
}

// ── Compose context detection ─────────────────────────────────────────────────

async function getComposeContext() {
  if (isMock()) return null
  const docker   = getDockerInstance()
  const hostname = os.hostname()
  const ctrs     = await docker.listContainers({ all: false })
  const self     = ctrs.find(c => c.Id.startsWith(hostname))
              || ctrs.find(c => c.Names.some(n => /node.?api/i.test(n)))
  if (!self) return null

  const L          = self.Labels || {}
  const workingDir = L["com.docker.compose.project.working_dir"]
  const configRaw  = L["com.docker.compose.project.config_files"] || ""

  if (!workingDir) return null

  // Build -f flags, mapping host paths → /workspace inside the helper container
  const flags = configRaw
    .split(",")
    .map(f => f.trim())
    .filter(Boolean)
    .map(f => `-f "${f.replace(workingDir, "/workspace")}"`)
    .join(" ")

  return { workingDir, flags }
}

// ── Update ────────────────────────────────────────────────────────────────────

// Simple in-memory update state (lost on container restart — intentional)
const updateState = { running: false, log: [], error: null, startedAt: null }

function getUpdateState() { return { ...updateState, log: [...updateState.log] } }

async function triggerUpdate() {
  if (isMock())         throw new Error("Update not available in mock mode")
  if (updateState.running) throw new Error("Update already in progress")

  const ctx = await getComposeContext()
  if (!ctx) throw new Error("Could not detect compose project directory — is this running inside Docker Compose?")

  const { workingDir, flags } = ctx

  // Build the shell script that will run inside the helper container
  const script = [
    "set -e",
    "apk add --no-cache git > /dev/null 2>&1",
    "cd /workspace",
    "echo '::pull:: Pulling latest code...'",
    "git pull",
    "echo '::down:: Stopping current services...'",
    `docker compose ${flags} down`,
    "echo '::build:: Building and starting updated services...'",
    `docker compose ${flags} up --build -d`,
    "echo '::done:: Update complete'",
  ].join("\n")

  // Reset state
  updateState.running   = true
  updateState.log       = ["Starting update helper container..."]
  updateState.error     = null
  updateState.startedAt = new Date().toISOString()

  const docker = getDockerInstance()

  try {
    const helper = await docker.createContainer({
      Image: "docker:cli",
      Cmd:   ["sh", "-c", script],
      HostConfig: {
        Binds: [
          "/var/run/docker.sock:/var/run/docker.sock",
          `${workingDir}:/workspace`,
        ],
        AutoRemove: true,
      },
    })

    // Stream helper container output into updateState.log
    const stream = await helper.attach({ stream: true, stdout: true, stderr: true })
    docker.modem.demuxStream(stream, {
      write(chunk) {
        const line = chunk.toString().trim()
        if (line) updateState.log.push(line)
      }
    }, {
      write(chunk) {
        const line = chunk.toString().trim()
        if (line) updateState.log.push("⚠ " + line)
      }
    })

    await helper.start()

    // Wait for helper to finish (up to 10 minutes)
    await new Promise((resolve, reject) => {
      helper.wait((err, data) => {
        if (err) return reject(err)
        if (data?.StatusCode !== 0) return reject(new Error(`Helper exited with code ${data?.StatusCode}`))
        resolve()
      })
    })

    updateState.running = false
  } catch (err) {
    updateState.running = false
    updateState.error   = err.message
    updateState.log.push("✕ " + err.message)
    throw err
  }
}

module.exports = { getVersionInfo, getUpdateState, triggerUpdate }

File: go-api/handlers/system.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"pulsenode/api/docker"
)

const repo = "SakithaSamarathunga33/vps"

type SystemHandler struct {
	dc          *docker.Client
	mu          sync.Mutex
	versionCache map[string]interface{}
	versionAt   time.Time
	updateState  updateStatus
}

type updateStatus struct {
	Running   bool     `json:"running"`
	Log       []string `json:"log"`
	Error     string   `json:"error,omitempty"`
	StartedAt string   `json:"startedAt,omitempty"`
}

func NewSystemHandler(dc *docker.Client) *SystemHandler {
	return &SystemHandler{dc: dc}
}

func (h *SystemHandler) Version(w http.ResponseWriter, r *http.Request) {
	h.mu.Lock()
	if h.versionCache != nil && time.Since(h.versionAt) < time.Hour {
		cached := h.versionCache
		h.mu.Unlock()
		writeJSON(w, 200, cached)
		return
	}
	h.mu.Unlock()

	current := currentVersion()
	info := map[string]interface{}{
		"current":    current,
		"latest":     nil,
		"hasUpdate":  false,
		"releaseUrl": nil,
		"changelog":  nil,
	}

	client := &http.Client{Timeout: 6 * time.Second}
	req, _ := http.NewRequest(http.MethodGet,
		"https://api.github.com/repos/"+repo+"/releases/latest", nil)
	req.Header.Set("Accept", "application/vnd.github.v3+json")
	req.Header.Set("User-Agent", "PulseNode")
	if resp, err := client.Do(req); err == nil {
		defer resp.Body.Close()
		var d struct {
			TagName string `json:"tag_name"`
			HTMLURL string `json:"html_url"`
			Body    string `json:"body"`
		}
		if json.NewDecoder(resp.Body).Decode(&d) == nil {
			latest := strings.TrimPrefix(d.TagName, "v")
			info["latest"] = latest
			info["hasUpdate"] = isNewer(latest, current)
			info["releaseUrl"] = d.HTMLURL
			info["changelog"] = d.Body
		}
	}

	h.mu.Lock()
	h.versionCache = info
	h.versionAt = time.Now()
	h.mu.Unlock()
	writeJSON(w, 200, info)
}

func (h *SystemHandler) UpdateStatus(w http.ResponseWriter, r *http.Request) {
	h.mu.Lock()
	st := h.updateState
	h.mu.Unlock()
	writeJSON(w, 200, st)
}

func (h *SystemHandler) Update(w http.ResponseWriter, r *http.Request) {
	if h.dc.IsMock() {
		writeErr(w, 503, "update not available in mock mode")
		return
	}
	h.mu.Lock()
	if h.updateState.Running {
		h.mu.Unlock()
		writeErr(w, 409, "update already in progress")
		return
	}
	h.mu.Unlock()

	go h.triggerUpdate()
	writeJSON(w, 200, map[string]interface{}{
		"started": true,
		"message": "Update started. The dashboard will restart shortly.",
	})
}

func (h *SystemHandler) triggerUpdate() {
	ctx := &contextDetect{dc: h.dc}
	workingDir, flags, err := ctx.detect()
	if err != nil {
		h.mu.Lock()
		h.updateState.Error = err.Error()
		h.mu.Unlock()
		return
	}

	script := strings.Join([]string{
		"set -e",
		"apk add --no-cache git > /dev/null 2>&1",
		"cd /workspace",
		"echo '::pull:: Pulling latest code...'",
		"git pull",
		"echo '::down:: Stopping current services...'",
		"docker compose " + flags + " down",
		"echo '::build:: Building and starting updated services...'",
		"docker compose " + flags + " up --build -d",
		"echo '::done:: Update complete'",
	}, "\n")

	h.mu.Lock()
	h.updateState = updateStatus{
		Running:   true,
		Log:       []string{"Starting update helper container..."},
		StartedAt: time.Now().UTC().Format(time.RFC3339),
	}
	h.mu.Unlock()

	cfg := docker.CreateContainerConfig{
		Image: "docker:cli",
		Cmd:   []string{"sh", "-c", script},
		HostConfig: docker.CreateHostConfig{
			Binds: []string{
				"/var/run/docker.sock:/var/run/docker.sock",
				workingDir + ":/workspace",
			},
		},
	}
	cid, err := h.dc.CreateContainer(cfg)
	if err != nil {
		h.mu.Lock()
		h.updateState.Running = false
		h.updateState.Error = err.Error()
		h.mu.Unlock()
		return
	}
	if err := h.dc.StartContainer(cid); err != nil {
		h.mu.Lock()
		h.updateState.Running = false
		h.updateState.Error = err.Error()
		h.mu.Unlock()
		return
	}

	// Poll container logs and wait for completion
	for {
		time.Sleep(2 * time.Second)
		logs, _ := h.dc.ContainerLogs(cid, 20)
		if logs != "" {
			h.mu.Lock()
			h.updateState.Log = append(h.updateState.Log, strings.Split(logs, "\n")...)
			h.mu.Unlock()
		}
		// Check if container finished
		ctrs, _ := h.dc.ListContainers(true)
		finished := true
		for _, c := range ctrs {
			if c.ID == cid {
				if c.State == "running" {
					finished = false
				}
				break
			}
		}
		if finished {
			break
		}
	}

	h.mu.Lock()
	h.updateState.Running = false
	h.mu.Unlock()
}

type contextDetect struct{ dc *docker.Client }

func (c *contextDetect) detect() (workingDir, flags string, err error) {
	ctrs, err := c.dc.ListContainers(false)
	if err != nil {
		return "", "", err
	}
	hostname, _ := os.Hostname()
	var self *docker.Container
	for i := range ctrs {
		if strings.HasPrefix(ctrs[i].ID, hostname) {
			self = &ctrs[i]
			break
		}
	}
	if self == nil {
		for i := range ctrs {
			for _, n := range ctrs[i].Names {
				if strings.Contains(strings.ToLower(n), "go-api") ||
					strings.Contains(strings.ToLower(n), "pulsenode") {
					self = &ctrs[i]
					break
				}
			}
		}
	}
	if self == nil {
		return "", "", fmt.Errorf("could not detect compose project container")
	}
	workingDir = self.Labels["com.docker.compose.project.working_dir"]
	if workingDir == "" {
		return "", "", fmt.Errorf("could not detect compose working directory")
	}
	configRaw := self.Labels["com.docker.compose.project.config_files"]
	var fflags []string
	for _, f := range strings.Split(configRaw, ",") {
		f = strings.TrimSpace(f)
		if f != "" {
			fflags = append(fflags, `-f "`+strings.ReplaceAll(f, workingDir, "/workspace")+`"`)
		}
	}
	flags = strings.Join(fflags, " ")
	return workingDir, flags, nil
}

func currentVersion() string {
	if v := os.Getenv("PULSENODE_VERSION"); v != "" {
		return v
	}
	return "dev"
}

func isNewer(a, b string) bool {
	pa := splitVer(a)
	pb := splitVer(b)
	for i := 0; i < max2(len(pa), len(pb)); i++ {
		va, vb := 0, 0
		if i < len(pa) {
			va, _ = strconv.Atoi(pa[i])
		}
		if i < len(pb) {
			vb, _ = strconv.Atoi(pb[i])
		}
		if va > vb {
			return true
		}
		if va < vb {
			return false
		}
	}
	return false
}

func splitVer(v string) []string { return strings.Split(v, ".") }
func max2(a, b int) int {
	if a > b {
		return a
	}
	return b
}
